aws/resource_aws_emr_instance_fleet.go

```diff
@@ -0,0 +1,369 @@
+									},
+									"timeout_duration_minutes": {
+										Type:     schema.TypeInt,
+										ForceNew: true,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"target_on_demand_capacity": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  0,
+			},
+			"target_spot_capacity": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  0,
+			},
+			"provisioned_on_demand_capacity": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"provisioned_spot_capacity": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsEMRInstanceFleetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+
+	addInstanceFleetInput := &emr.AddInstanceFleetInput{
+		ClusterId: aws.String(d.Get("cluster_id").(string)),
+	}
+
+	taskFleet := map[string]interface{}{
+		"name":                      d.Get("name"),
+		"target_on_demand_capacity": d.Get("target_on_demand_capacity"),
+		"target_spot_capacity":      d.Get("target_spot_capacity"),
+		"instance_type_configs":     d.Get("instance_type_configs"),
+		"launch_specifications":     d.Get("launch_specifications"),
+	}
+	addInstanceFleetInput.InstanceFleet = readInstanceFleetConfig(taskFleet, emr.InstanceFleetTypeTask)
+
+	log.Printf("[DEBUG] Creating EMR instance fleet params: %s", addInstanceFleetInput)
+	resp, err := conn.AddInstanceFleet(addInstanceFleetInput)
+	if err != nil {
+		return fmt.Errorf("error adding EMR Instance Fleet: %w", err)
+	}
+
+	log.Printf("[DEBUG] Created EMR instance fleet finished: %#v", resp)
+	if resp == nil {
+		return fmt.Errorf("error creating instance fleet: no instance fleet returned")
+	}
+	d.SetId(*resp.InstanceFleetId)
+
+	return nil
+}
+
+func resourceAwsEMRInstanceFleetRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+	instanceFleets, err := fetchAllEMRInstanceFleets(conn, d.Get("cluster_id").(string))
+	if err != nil {
+		log.Printf("[DEBUG] EMR doesn't have any Instance Fleet ")
+		d.SetId("")
+		return nil
+	}
+
+	fleet := findInstanceFleetById(instanceFleets, d.Id())
+	if fleet == nil {
+		log.Printf("[DEBUG] EMR Instance Fleet (%s) not found, removing", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	taskFleet := flattenInstanceFleet(fleet)[0].(map[string]interface{})
+	d.Set("name", taskFleet["name"])
+	d.Set("target_on_demand_capacity", taskFleet["target_on_demand_capacity"])
+	d.Set("target_spot_capacity", taskFleet["target_spot_capacity"])
+	d.Set("instance_type_configs", taskFleet["instance_type_configs"])
+	d.Set("launch_specifications", taskFleet["launch_specifications"])
+	d.Set("provisioned_on_demand_capacity", taskFleet["provisioned_on_demand_capacity"])
+	d.Set("provisioned_spot_capacity", taskFleet["provisioned_spot_capacity"])
```

Comment:
We can skip the indirection here by directly getting what we need:

```suggestion
	if err := d.Set("instance_type_configs", flatteninstanceTypeConfigs(fleet.InstanceTypeSpecifications)); err != nil
		return fmt.Errorf("error setting instance_type_configs: %w", err)
	}

	if err := d.Set("launch_specifications", flattenLaunchSpecifications(fleet.LaunchSpecifications); err != nil
		return fmt.Errorf("error setting launch_specifications: %w", err)
	}

	d.Set("name", fleet.Name)
	d.Set("provisioned_on_demand_capacity", fleet.ProvisionedOnDemandCapacity)
	d.Set("provisioned_spot_capacity", fleet.ProvisionedSpotCapacity)
	d.Set("target_on_demand_capacity", fleet.TargetOnDemandCapacity)
	d.Set("target_spot_capacity", fleet.TargetSpotCapacity)
```